c10/core/DispatchKeySet.cpp

```diff
@@ -2,16 +2,73 @@
 
 namespace c10 {
 
-static DispatchKeySet autograd_dispatch_keys{
-  DispatchKey::Autograd,
-  DispatchKey::AutogradXLA,
-  DispatchKey::PrivateUse1_PreAutograd,
-  DispatchKey::PrivateUse2_PreAutograd,
-  DispatchKey::PrivateUse3_PreAutograd,
-};
-
-DispatchKeySet AutogradDispatchKeys() {
-  return autograd_dispatch_keys;
+// [Note] Static initialize alias -> runtime key mapping.
+// Construct on the first use since it's called when we do TorchLibraryInit
+// We have to use a raw pointer here to make sure it's still available when
+// we destruct the dispatch table.
+// http://www.cs.technion.ac.il/users/yechiel/c++-faq/construct-on-first-use-v2.html
+static DispatchKeySet& autograd_dispatch_keys (void) {
+  static DispatchKeySet* autograd_keys = new DispatchKeySet({
```

Comment:
I think we can just define this as constexpr and then we don't have to worry about static initialization at all. If the initializer list DispatchKeySet constructor is not constexpr, it should be made constexpr.

Reply:
(The destructor order business is nasty and I didn't realize it would be a problem, sorry!)